app/models/container_deployment.rb

```diff
@@ -0,0 +1,313 @@
+  has_one :deployment_authentication
+  belongs_to :deployed_ext_management_system, :class_name => 'ExtManagementSystem'
+  belongs_to :deployed_on_ext_management_system, :class_name => 'ExtManagementSystem'
+  belongs_to :automation_task
+  has_many :container_deployment_nodes
+
+  DEPLOYMENT_TYPES = ['OpenShift Origin', 'OpenShift Enterprise', 'Atomic Enterprise'].freeze
+
+  def self.supported_types
+    DEPLOYMENT_TYPES
+  end
+
+  def masters
+    container_deployment_nodes.find_tagged_with(:any => "/user/master", :ns => "*")
+  end
+
+  def deployment_master
+    container_deployment_nodes.find_tagged_with(:any => "/user/deployment_master", :ns => "*").first
+  end
+
+  def nodes
+    container_deployment_nodes.find_tagged_with(:any => "/user/node", :ns => "*")
+  end
+
+  def nfs_ips_or_hostnames
+    to_ip_or_hostname_array container_deployment_nodes.find_tagged_with(:any => "/user/nfs", :ns => "*")
+  end
+
+  def masters_ips_or_hostnames
+    to_ip_or_hostname_array masters
+  end
+
+  def deployment_master_ips_or_hostnames
+    extract_public_ip_or_hostname deployment_master
+  end
+
+  def nodes_ips_or_hostnames
+    to_ip_or_hostname_array nodes
+  end
+
+  def add_deployment_provider(options)
+    provider = ExtManagementSystem.new(
+      :type      => options[:provider_type],
+      :name      => options[:provider_name],
+      :port      => options[:provider_port],
+      :hostname  => options[:provider_hostname],
+      :ipaddress => options[:provider_ipaddress])
+    provider.save!
+    provider.update_authentication(:bearer => {:auth_key => options[:auth_key], :save => true})
+    valid_provider = provider.authentication_check.first
+    if valid_provider
+      self.deployed_ext_management_system = provider
+      save!
+    else
+      provider.destroy
+    end
+    valid_provider
+  end
+
+  def assign_container_deployment_node(options)
+    send(options[:type]).each do |deploymnet_node|
+      next unless deploymnet_node.vm.nil?
+      deploymnet_node.vm = Vm.find(options[:vm_id])
+      deploymnet_node.save!
+    end
+  end
+
+  def to_ip_or_hostname_array(container_deployment_nodes)
+    ip_array = []
+    container_deployment_nodes.each do |container_deployment_node|
+      ip_array << if container_deployment_node.vm
+                    container_deployment_node.vm.hardware.ipaddresses.last
+                  else
+                    container_deployment_node.ip_or_hostname
+                  end
+    end
+    ip_array
+  end
+
+  def generate_ansible_inventory_for_subscription
+    template = "
+[OSEv3:children]
+masters
+nodes
+
+[OSEv3:vars]
+ansible_ssh_user=#{ssh_user}
+deployment_type=openshift-enterprise
+rhsub_user=#{rhn_user}
+rhsub_pass=#{rhn_pass}
+rhsub_pool=#{rhn_sku}
+
+[masters]
+#{masters_ips_or_hostnames.join("\n") unless masters.empty?}
+[nodes]
+#{nodes_ips_or_hostnames.join("\n") unless nodes.empty?}
+    "
+    template
+  end
+
+  def generate_ansible_inventory
+    byebug
+    template = "
```

Comment:
You should [heredoc](http://rubyquicktips.com/post/4438542511/heredoc-and-indent) this.  You should still be able to do variable interpolation in heredocs, I believe.

But there might be a better way to externalize this.  I know that the SCVMM (Microsoft InfraManager) provider has Powershell scripts in its directory.  You might do something like that instead to pull this out.